From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 8 Jan 2026 04:31:53 -0500
Subject: [PATCH] Configurable vanilla profiler

License: MIT (https://opensource.org/licenses/MIT)
Gale - https://galemc.org

Co-authored-by: Martijn Muijsers <martijnmuijsers@live.nl>

This patch is based on the following patch:
"Remove Mojang Profiler"
By: BillyGalbreath <blake.galbreath@gmail.com>
As part of: Purpur (https://github.com/PurpurMC/Purpur)
Licensed under: MIT (https://opensource.org/licenses/MIT)

* Purpur copyright *

MIT License

Copyright (c) 2019-2022 PurpurMC

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index b56481491db6e39ca30c482ce1d6cd5d44d65d78..4cbbcb757ce1ee065a49539d9237d68b5fd4dde8 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1355,7 +1355,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
                 // Paper - improve tick loop - done above
 
-                try (Profiler.Scope ignored = Profiler.use(this.createProfiler())) {
+                try (Profiler.Scope ignored = org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_PROFILER ? null : Profiler.use(this.createProfiler())) { // Leaf - Configurable vanilla profiler
                     this.processPacketsAndTick(sprinting);
                     ProfilerFiller profiler = Profiler.get();
                     profiler.push("nextTickWait");
diff --git a/net/minecraft/server/ReloadableServerResources.java b/net/minecraft/server/ReloadableServerResources.java
index 4b7cb2ba2e1f05d73768e62bb93cd129514166d0..49c1b41143b99da4d72ff8e239b99c6c047132ba 100644
--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -112,7 +112,7 @@ public class ReloadableServerResources {
                                     backgroundExecutor,
                                     mainThreadExecutor,
                                     DATA_RELOAD_INITIAL_TASK,
-                                    LOGGER.isDebugEnabled()
+                                    !org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_PROFILER // Leaf - Configurable vanilla profiler
                                 )
                                 .done()
                                 .thenApply(ignore -> result);
diff --git a/net/minecraft/server/commands/DebugCommand.java b/net/minecraft/server/commands/DebugCommand.java
index 812659bc14a356bde510d245557f94746bc576cc..f665f112ab278c786403d98fc206d6752fd4de1f 100644
--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -71,6 +71,7 @@ public class DebugCommand {
     }
 
     private static int start(final CommandSourceStack source) throws CommandSyntaxException {
+        if (showRemovalWarning(source, "debug start")) return 0; // Leaf - Configurable vanilla profiler
         MinecraftServer server = source.getServer();
         if (server.isTimeProfilerRunning()) {
             throw ERROR_ALREADY_RUNNING.create();
@@ -82,6 +83,7 @@ public class DebugCommand {
     }
 
     private static int stop(final CommandSourceStack source) throws CommandSyntaxException {
+        if (showRemovalWarning(source, "debug stop")) return 0; // Leaf - Configurable vanilla profiler
         MinecraftServer server = source.getServer();
         if (!server.isTimeProfilerRunning()) {
             throw ERROR_NOT_RUNNING.create();
@@ -99,6 +101,21 @@ public class DebugCommand {
         return (int)tps;
     }
 
+    // Leaf start - Configurable vanilla profiler
+    public static boolean showRemovalWarning(final CommandSourceStack source, final String command) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_PROFILER) {
+            net.kyori.adventure.text.minimessage.MiniMessage mm = net.kyori.adventure.text.minimessage.MiniMessage.miniMessage();
+
+            source.getSender().sendMessage(mm.deserialize("<gold>/" + command + " command is unavailable, since Minecraft's vanilla profiler has been disabled."));
+            source.getSender().sendMessage(mm.deserialize("<gold>Please use <click:suggest_command:'/spark'><grey>/spark</grey></click> instead, or check Leaf docs for more information:"));
+            source.getSender().sendMessage(mm.deserialize("<gold><click:open_url:'" + org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_PROFILER_DOCS_URL + "'>" + org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_PROFILER_DOCS_URL + "</click>"));
+
+            return true;
+        }
+        return false;
+    }
+    // Leaf end - Configurable vanilla profiler
+
     private static class TraceCustomExecutor
         extends CustomCommandExecutor.WithErrorHandling<CommandSourceStack>
         implements CustomCommandExecutor.CommandAdapter<CommandSourceStack> {
diff --git a/net/minecraft/server/commands/PerfCommand.java b/net/minecraft/server/commands/PerfCommand.java
index f881d6c20487569c0408fbb33e5d9d09363d7cbc..767c812271672271123446265b6ee75f2725ae88 100644
--- a/net/minecraft/server/commands/PerfCommand.java
+++ b/net/minecraft/server/commands/PerfCommand.java
@@ -42,6 +42,7 @@ public class PerfCommand {
     }
 
     private static int startProfilingDedicatedServer(final CommandSourceStack source) throws CommandSyntaxException {
+        if (net.minecraft.server.commands.DebugCommand.showRemovalWarning(source, "perf start")) return 0; // Leaf - Configurable vanilla profiler
         MinecraftServer server = source.getServer();
         if (server.isRecordingMetrics()) {
             throw ERROR_ALREADY_RUNNING.create();
@@ -55,6 +56,7 @@ public class PerfCommand {
     }
 
     private static int stopProfilingDedicatedServer(final CommandSourceStack source) throws CommandSyntaxException {
+        if (net.minecraft.server.commands.DebugCommand.showRemovalWarning(source, "perf stop")) return 0; // Leaf - Configurable vanilla profiler
         MinecraftServer server = source.getServer();
         if (!server.isRecordingMetrics()) {
             throw ERROR_NOT_RUNNING.create();
diff --git a/net/minecraft/server/packs/resources/ReloadableResourceManager.java b/net/minecraft/server/packs/resources/ReloadableResourceManager.java
index 2bccd6a2ade7e5d2bda6f031202700f9811b8f01..222c7d3c88fba00eaead059425e2be2a7983d5e1 100644
--- a/net/minecraft/server/packs/resources/ReloadableResourceManager.java
+++ b/net/minecraft/server/packs/resources/ReloadableResourceManager.java
@@ -45,7 +45,7 @@ public class ReloadableResourceManager implements AutoCloseable, ResourceManager
         LOGGER.info("Reloading ResourceManager: {}", LogUtils.defer(() -> resourcePacks.stream().map(PackResources::packId).collect(Collectors.joining(", "))));
         this.resources.close();
         this.resources = new MultiPackResourceManager(this.type, resourcePacks);
-        return SimpleReloadInstance.create(this.resources, this.listeners, backgroundExecutor, mainThreadExecutor, initialTask, LOGGER.isDebugEnabled());
+        return SimpleReloadInstance.create(this.resources, this.listeners, backgroundExecutor, mainThreadExecutor, initialTask, !org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_PROFILER); // Leaf - Configurable vanilla profiler
     }
 
     @Override
diff --git a/net/minecraft/util/profiling/Profiler.java b/net/minecraft/util/profiling/Profiler.java
index ae5974d304abd7d240453f0376c5fd79359a9585..a7ae88f744e1adafca1e13243dbee94536c9c961 100644
--- a/net/minecraft/util/profiling/Profiler.java
+++ b/net/minecraft/util/profiling/Profiler.java
@@ -49,6 +49,7 @@ public final class Profiler {
     }
 
     public static ProfilerFiller get() {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_PROFILER) return InactiveProfiler.INSTANCE; // Leaf - Configurable vanilla profiler
         return ACTIVE_COUNT.get() == 0 ? getDefaultFiller() : Objects.requireNonNullElseGet(ACTIVE.get(), Profiler::getDefaultFiller);
     }
 
